internal/manifestwork/manifestwork.go

```diff
@@ -175,3 +194,92 @@ func (mwg *manifestWorkGenerator) getOwnedManifestWorks(ctx context.Context, mcm
 
 	return manifestWorkList, err
 }
+
+func (mwg *manifestWorkGenerator) managedClusterKernelMappings(
+	ctx context.Context,
+	mcm hubv1beta1.ManagedClusterModule,
+	kernelVersions []string) []kmmv1beta1.KernelMapping {
+
+	mod := &kmmv1beta1.Module{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      mcm.Name,
+			Namespace: mcm.Spec.SpokeNamespace,
+		},
+		Spec: mcm.Spec.ModuleSpec,
+	}
+
+	logger := log.FromContext(ctx)
+
+	mappings := make([]kmmv1beta1.KernelMapping, 0, len(kernelVersions))
+
+	for _, kernelVersion := range kernelVersions {
+		kernelVersion := strings.TrimSuffix(kernelVersion, "+")
+
+		kernelVersionLogger := logger.WithValues(
+			"kernel version", kernelVersion,
+		)
+
+		mld, err := mwg.kernelAPI.GetModuleLoaderDataForKernel(mod, kernelVersion)
+		if err != nil {
+			kernelVersionLogger.Info("no suitable container image found; skipping kernel version")
+			continue
+		}
+
+		kernelVersionLogger.V(1).Info("Found a valid mapping",
+			"image", mld.ContainerImage,
+		)
+
+		mapping := mwg.mappingFromModuleLoaderData(mld)
+
+		digest, err := mwg.imageDigestForModuleLoaderData(ctx, mld)
+		if err != nil {
+			kernelVersionLogger.Error(err, "skipping image tag replacement")
+		}
+		if digest != "" {
```

Comment:
What happens if `digest == ""`? We keep the tag?

Reply:
I don't have a strong opinion about it but I am just wondering if it won't be confusing why some of the images were manipulated while other where skipped. I can see pros and cons for both approaches (returning an error on the hub or just letting the spoke fail) so please proceed with the option that you see best fits.